### PR TITLE
BUG: Stringify DS attrs in dir

### PR DESCRIPTION
**Pull request by MaximilianR**
Previously:

```python
In [21]: ds
Out[21]: 
<xarray.Dataset>
Dimensions:  (abc: 3)
Coordinates:
  * abc      (abc) int64 0 1 2
Data variables:
    0        (abc) float64 0.3168 0.1132 0.6664
    1        (abc) float64 0.9799 0.9663 0.01084
    2        (abc) float64 0.9171 0.6848 0.5028


In [22]: dir(ds)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-22-fd55ce5c03db> in <module>()
----> 1 dir(ds)

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xarray/core/common.py in __dir__(self)
    157         extra_attrs = [item for sublist in self._attr_sources
    158                        for item in sublist]
--> 159         return sorted(set(dir(type(self)) + extra_attrs))
    160 
    161 

TypeError: unorderable types: numpy.ndarray() > str()
```

### CODE REVIEW

**Comment by MaximilianR** [comment]
AppVeyor break is an existing issue I think?

File: xarray/test/test_dataset.py [hunk 1]
```diff
@@ -323,6 +323,16 @@ def test_attr_access(self):
         self.assertEqual(ds.attrs['tmin'], -999)
         self.assertDataArrayIdentical(ds.tmin, ds['tmin'])
 
+    def test_dir(self):
+        ds = create_test_data()
+        # add a numbered key to ensure this doesn't break dir
+        ds[5] = list(ds.keys())[0]
+
+        some_expected_attrs = {'pipe', 'mean', 'isnull', 'var1',
+                               'dim1', 'numbers'}
```
Comment (jhamman) [inline xarray/test/test_dataset.py]: this breaks the Python 2.6 test on travis.  Use a list or a tuple.

Comment (MaximilianR) [inline xarray/test/test_dataset.py]: We still support 2.6? How about dropping that...

Comment (shoyer) [inline xarray/test/test_dataset.py]: The plan (per mailing list discussino) is to drop Python 2.6 support for the next major release. That should probably be the next release, but we'll need to cut it out of the Travis-CI setup first.

File: xarray/test/test_dataset.py [hunk 2]
```diff
@@ -323,6 +323,16 @@ def test_attr_access(self):
         self.assertEqual(ds.attrs['tmin'], -999)
         self.assertDataArrayIdentical(ds.tmin, ds['tmin'])
 
+    def test_dir(self):
+        ds = create_test_data()
+        # add a numbered key to ensure this doesn't break dir
+        ds[5] = list(ds.keys())[0]
```
Comment (shoyer) [inline xarray/test/test_dataset.py]: Can you simply create the key directly, e.g., `ds[5] = 'foo'`? It's not clear to me why you're using indexing here.

**Comment by shoyer** [comment]
this looks good, once you fix a few quick nits

File: xarray/test/test_dataset.py (hunk 1 above)
Comment (jhamman) [inline xarray/test/test_dataset.py]: see #855 and #856 for issue/PR on removing Python 2.6 support.

**Comment by shoyer** [comment]
@MaximilianR Can you rebase this? Otherwise this looks good to go to me.

File: doc/whats-new.rst [hunk 3]
```diff
@@ -51,6 +51,10 @@ Bug fixes
    This fixes issue :issue:`665`.
    `Filipe Fernandes <https://github.com/ocefpaf>`_.
 
+- ``Dataset.__dir__()`` (the method python calls to get autocomplete options) failed
+  if one of the ``DataSet``'s keys was not a string (:issue:`852`).
```
Comment (shoyer) [inline doc/whats-new.rst]: don't forget to give yourself credit :)

**Comment by MaximilianR** [comment]
I've been remiss about the final fixes here. Will need a couple of weeks on other commitments but I will get to this

**Comment by MaximilianR** [comment]
@shoyer @jhamman updated, but travis queue looks stopped?

File: doc/whats-new.rst [hunk 4]
```diff
@@ -24,6 +24,11 @@ Bug fixes
 - Fix regression in v0.8.0 that broke assignment to Datasets with non-unique
   indexes (:issue:`943`).
 
+- ``Dataset.__dir__()`` (i.e. the method python calls to get autocomplete options) failed
+  if one of the ``DataSet``'s keys was not a string (:issue:`852`). By
```
Comment (jhamman) [inline doc/whats-new.rst]: `DataSet` --> `Dataset`

**Comment by MaximilianR** [comment]
Rebased @shoyer @jhamman

**Comment by shoyer** [comment]
Looking at this again -- maybe better just to drop non-strings from `__dir__`?

For example, if you put `None` is an `xr.Dataset`, this would give you autocomplete for `ds.None`, which won't actually work when it's mapped to `ds['None']`.

**Comment by MaximilianR** [comment]
> Looking at this again -- maybe better just to drop non-strings from __dir__?

Yes!

File: xarray/core/common.py [hunk 5]
```diff
@@ -212,7 +212,7 @@ def __dir__(self):
         methods.
         """
         extra_attrs = [item for sublist in self._attr_sources
-                       for item in sublist]
+                       for item in sublist if isinstance(item, str)]
```
Comment (shoyer) [inline xarray/core/common.py]: This should be `pycompat.basestring` -- otherwise I think this would break for unicode variable names.

Comment (MaximilianR) [inline xarray/core/common.py]: Great. Not putting `bytes` there as I get `E       TypeError: unorderable types: bytes() < str()`. Let me know if that's not OK though

Comment (shoyer) [inline xarray/core/common.py]: basestring is a super class of both bytes and str on Python 2. On Python 3, it's only str.

**Comment by shoyer** [comment]
Merged in 15b4291bd9e02440a5bd5346b55a8f6bee939690

File: doc/whats-new.rst [hunk 6]
```diff
@@ -26,6 +26,10 @@ Bug fixes
 
 - Fix to ensure xarray works with h5netcdf v0.3.0 for arrays with ``dtype=str``
   (:issue:`953`). By `Stephan Hoyer <https://github.com/shoyer>`_.
+- ``Dataset.__dir__()`` (i.e. the method python calls to get autocomplete options) failed
+  if one of the ``Dataset``'s keys was not a string (:issue:`852`). By
```
Comment (shoyer) [inline doc/whats-new.rst]: Just a note, in restructured text this breaks:

    ``Dataset``'s

You need whitespace or punctuation after the backticks.

**Pull request closed without merging**